src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,224 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            Node<T> node = new Node<>(value);
+            node.prev = null;
+            node.next = null;
+            head = node;
+            tail = node;
+            size++;
+        } else {
+            addNode(value);
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        findIndex(index);
+        if (index == size && index != 0) {
+            addNode(value);
+        } else if (index == 0) {
+            addFirst(value);
+        } else if (size / 2 >= index) {
+            addFromHead(value, index);
+        } else {
+            addFromTail(value, index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        findIndexGet(index);
+        return size / 2 >= index ? getFromHead(index) : getFromTail(index);
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        findIndexGet(index);
+        return size / 2 >= index ? setFromHead(value, index) : setFromTail(value, index);
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        findIndexGet(index);
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        if (current == head) {
+            head = current.next;
+        } else {
+            current.prev.next = current.next;
+        }
+        if (current == tail) {
+            tail = current.prev;
+        } else {
+            current.next.prev = current.prev;
+        }
```

Comment:
in common mistakes, it is advised to move this logic to a separate method, because it is also used in the next remove method

Reply:
Agree create method unlink (see common mistakes)